L82-core-event-engine.md

```diff
@@ -0,0 +1,564 @@
+```
+
+### DNS Resolution
+
+```cpp
+  /// The DNSResolver that provides asynchronous resolution.
+  class DNSResolver {
+   public:
+    /// A task handle for DNS Resolution requests.
+    struct LookupTaskHandle {
+      intptr_t key;
+    };
+    /// A DNS SRV record type.
+    struct SRVRecord {
+      std::string host;
+      int port = 0;
+      int priority = 0;
+      int weight = 0;
+    };
+    /// Called with the collection of sockaddrs that were resolved from a given
+    /// target address.
+    using LookupHostnameCallback =
+        std::function<void(absl::StatusOr<std::vector<ResolvedAddress>>)>;
+    /// Called with a collection of SRV records.
+    using LookupSRVCallback =
+        std::function<void(absl::StatusOr<std::vector<SRVRecord>>)>;
+    /// Called with the result of a TXT record lookup
+    using LookupTXTCallback = std::function<void(absl::StatusOr<std::string>)>;
+
+    virtual ~DNSResolver() = default;
+
+    /// Asynchronously resolve an address.
+    ///
+    /// \a default_port may be a non-numeric named service port, and will only
+    /// be used if \a address does not already contain a port component.
+    ///
+    /// When the lookup is complete, the \a on_resolve callback will be invoked
+    /// with a status indicating the success or failure of the lookup.
+    /// Implementations should pass the appropriate statuses to the callback.
+    /// For example, callbacks might expect to receive DEADLINE_EXCEEDED when
+    /// the deadline is exceeded or CANCELLED if the lookup was cancelled.
+    virtual LookupTaskHandle LookupHostname(LookupHostnameCallback on_resolve,
+                                            absl::string_view address,
+                                            absl::string_view default_port,
+                                            absl::Time deadline) = 0;
+    /// Asynchronously perform an SRV record lookup.
+    ///
+    /// \a on_resolve has the same meaning and expectations as \a
+    /// LookupHostname's \a on_resolve callback.
+    virtual LookupTaskHandle LookupSRV(LookupSRVCallback on_resolve,
+                                       absl::string_view name,
+                                       absl::Time deadline) = 0;
+    /// Asynchronously perform a TXT record lookup.
+    ///
+    /// \a on_resolve has the same meaning and expectations as \a
+    /// LookupHostname's \a on_resolve callback.
+    virtual LookupTaskHandle LookupTXT(LookupTXTCallback on_resolve,
+                                       absl::string_view name,
+                                       absl::Time deadline) = 0;
+    /// Cancel an asynchronous lookup operation.
+    virtual void TryCancelLookup(LookupTaskHandle handle) = 0;
+  };
+```
+
+### Task Execution
+
+```cpp
+  /// Retrieves an instance of a DNSResolver.
+  virtual absl::StatusOr<std::unique_ptr<DNSResolver>> GetDNSResolver() = 0;
+
+  /// Intended for future expansion of Task run functionality.
+  struct RunOptions {};
```

Comment:
What behavior should an EventEngine implementation have when it sees a RunOptions struct it does not understand?

How could we verify that RunOptions is understood by an implementation?

IOW: if I upgrade grpc and an option is added to this struct, how can calling code guarantee it's observed (this is the minimum bar to making a useful addition here), and how an EventEngine implementor know that the interface has expanded and some modifications are required?

Reply:
One proposal that could help here would be to add a `static constexpr int kVersion = 1;` and encourage implementations to take up a static_assert on the version, with documentation around why.

Reply:
> if I upgrade grpc and an option is added to this struct, how can calling code guarantee it's observed (this is the minimum bar to making a useful addition here)

I disagree. If gRPC ships with an implementation that gRPC internals all agree with, there's no question. If some custom EventEngine does not want to honor some set of RunOptions, I think that's entirely the implementer's prerogative. As I commented above with the `priority` example, if some impl does not distinguish between run priorities, I don't think that should be a show stopper, nor do I think gRPC internals would necessarily do something differently if that option were honored or not.

> How could we verify that RunOptions is understood by an implementation?

Can you give examples of when you think that would be meaningful?

> how [can] an EventEngine implementor know that the interface has expanded and some modifications are required?

Since this is a public API, we will file gRFCs for all EventEngine API changes. The mailing list will be notified as well.

Reply:
Regarding API changes: if we make a semantic addition here, grfcs and mailing list posts are likely insufficient. On a larger project: it's likely whomever did the integration work that required the custom event engine has moved on, and other folks are only loosely aware of the state required. If we add required semantics to RunOptions it's likely that ignoring them will work kind of but introduce subtle bugs, leading to all sorts of difficult to diagnose from a distance behaviors.

Which means that our rule for RunOptions must be, if we include it without a versioning scheme, that anything added must be documented thoroughly that it may be ignored at will by any and all EventEngine implementations, and we must train our team thoroughly that any usage of RunOptions could be ignored by customer code and to aware of that when reviewing code and when diagnosing support tickets.

As to a change that might cause problems... Consider: adding a 'same_thread' constraint to Run to ensure that whatever function is passed to Run is additionally run within the same thread, presumably to catch some thread safety optimization.

Reply:
Pinging @nicolasnoble for another opinion on this. I feel that if anything in `RunOptions` becomes essential to EventEngine functionality, that's a code smell. Backwards-incompatible changes should be first-class parts of the API, and merit gRPC version changes.

Reply:
I'd put forward a vote to just remove RunOptions from the API then: extending it seems too difficult to actually rely on, and any changes ought to merit a version change.

Reply:
After meeting about it, we concluded it makes sense to remove `RunOptions` and revisit the need for them before we promote this API from experimental status. In the meantime, we don't have any current use for the expansion, and we will otherwise be able to modify the API if we need to. I've removed `RunOptions` from the proposal.